cranelift/codegen/src/isa/aarch64/lower/isle.rs

```diff
@@ -206,24 +207,44 @@ impl Context for IsleContext<'_, '_, MInst, AArch64Backend> {
     ///
     /// The logic here is nontrivial enough that it's not really worth porting
     /// this over to ISLE.
-    fn load_constant64_full(
+    fn load_constant_full(
         &mut self,
         ty: Type,
         extend: &generated_code::ImmExtend,
+        extend_to: &OperandSize,
         value: u64,
     ) -> Reg {
         let bits = ty.bits();
-        let value = if bits < 64 {
-            if *extend == generated_code::ImmExtend::Sign {
-                let shift = 64 - bits;
-                let value = value as i64;
-
-                ((value << shift) >> shift) as u64
-            } else {
-                value & !(u64::MAX << bits)
-            }
-        } else {
-            value
+
+        let value = match extend_to {
+            OperandSize::Size32 => {
+                if bits < 32 {
+                    if *extend == generated_code::ImmExtend::Sign {
+                        let shift = 32 - bits;
+                        let value = value as i32;
+
+                        ((value << shift) >> shift) as u64
+                    } else {
+                        value & !((u32::MAX as u64) << bits)
+                    }
+                } else {
+                    value
+                }
+            },
+            OperandSize::Size64 => {
+                if bits < 64 {
+                    if *extend == generated_code::ImmExtend::Sign {
```

Comment:
Likewise here, and we can pull the `if bits < 64` in as a guard on one of the match arms as well.